drivers/mtd_flashpage/mtd_flashpage.c

```diff
@@ -81,8 +132,43 @@ static int _write(mtd_dev_t *dev, const void *buf, uint32_t addr, uint32_t size)
     return 0;
 }
 
+static int _write_page(mtd_dev_t *dev, const void *buf, uint32_t page, uint32_t offset,
+                       uint32_t size)
+{
+    mtd_flashpage_t *super = container_of(dev, mtd_flashpage_t, base);
+
+    page += super->offset;
+    uintptr_t addr = (uintptr_t)flashpage_addr(page) + offset;
+
+    DEBUG("flashpage: write %"PRIu32" bytes from %p to %p\n", size, buf, (void *)addr);
+
+    if ((addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT) ||
+        ((uintptr_t)buf % FLASHPAGE_WRITE_BLOCK_ALIGNMENT)) {
+        uint8_t tmp[FLASHPAGE_WRITE_BLOCK_SIZE];
+
+        offset = addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT;
+        size = MIN(size, FLASHPAGE_WRITE_BLOCK_ALIGNMENT - offset);
+
+        DEBUG("flashpage: write %"PRIu32" unaligned bytes\n", size);
+
+        memcpy(&tmp[0], (uint8_t *)addr - offset, sizeof(tmp));
+        memcpy(&tmp[offset], buf, size);
+
+        flashpage_write((uint8_t *)addr - offset, tmp, sizeof(tmp));
+
+        return size;
+    }
```

Comment:
Do we need support for unaligned writes? The `mtd` driver has the `write_size` field. It should be respected by `mtd_write_page_raw()` and `mtd_write()` users. IIRC `mtd_write_page()` takes care of alignment.

Actually, you're initializing `write_size` correctly in `MTD_FLASHPAGE_AUX_INIT_VAL`?

Or am I wrong? MTD is still confusing, I guess :D

Reply:
> It should be respected by `mtd_write_page_raw()` and `mtd_write()` users.

It's not so simple when that user is third party code, e.g. a file system and having this code duplicated for all users also doesn't sound very appealing. 
There is a case for moving this to the MTD layer proper though.

`mtd_write_page()` will take care of alignment only because it does a full read-modify-erase-write cycle, which is expensive.

Reply:
Do you have users of MTD that wouldn't respect `write_size`?

If yes, I'd rather add a mtd layer to align writes if needed instead of bringing this into every driver. IIRC `mtd_flashpage` doesn't' align in most cases. If seen this while working on `pkg_tinyvcdiff`.

If we align every driver to be able to address byte-wise: What was then the reason for introducing `write_size`?

Reply:
> Do you have users of MTD that wouldn't respect `write_size`?

At the moment, most of them - at least they are pretty much agnostic about it.
The filesystems all assume flash where you can write single bytes. `write_size` is currently not used for anything yet.

We need a separate buffer to align the writes, but for internal flash it's easy as we know the write size at compile time.